eZ/Bundle/EzPublishCoreBundle/Command/NormalizeImagesPathsCommand.php

```diff
@@ -209,9 +173,117 @@ private function moveFile(
             ]
         );
 
-        $newBinaryFile = $this->ioService->createBinaryFile($binaryCreateStruct);
-        if ($newBinaryFile instanceof BinaryFile) {
-            $this->ioService->deleteBinaryFile($oldBinaryFile);
+        // Before creating a new file validate if the same file doesn't exist already in order to not duplicate files
+        $newBinaryFile = $this->ioService->loadBinaryFileByUri(\DIRECTORY_SEPARATOR . $newPath);
+        if ($newBinaryFile instanceof MissingBinaryFile) {
+            $this->ioService->createBinaryFile($binaryCreateStruct);
+        }
+    }
+
+    protected function updateImagePathsToNormalize(
+        $imageData,
+        array $imagePathsToNormalize
+    ): array {
+        $filePath = $imageData['filepath'];
+        $fieldId = (int)$imageData['contentobject_attribute_id'];
+
+        $finalNormalizedPath = $this->getFinalNormalizedPath(
+            $filePath,
+            $imagePathsToNormalize
+        );
+
+        if ($finalNormalizedPath !== $filePath) {
+            $imagePathsToNormalize[$fieldId] = [
+                'fieldId' => $fieldId,
+                'oldPath' => $filePath,
+                'newPath' => $finalNormalizedPath,
+            ];
         }
+
+        return $imagePathsToNormalize;
+    }
+
+    protected function getFinalNormalizedPath(
+        string $filePath,
+        array $imagePathsToNormalize
+    ): string {
+        $processedPaths = array_values(
+            array_filter(
+                $imagePathsToNormalize,
+                static function (array $data) use ($filePath) {
+                    return $data['oldPath'] === $filePath;
+                }
+            )
+        );
+
+        return !empty($processedPaths)
+            ? $processedPaths[0]['newPath']
+            : $this->filePathNormalizer->normalizePath($filePath);
+    }
+
+    protected function getImagePathsToNormalize(SymfonyStyle $io): array
+    {
+        $imagesCount = $this->imageGateway->countDistinctImages();
+        $imagePathsToNormalize = [];
+        $iterations = ceil($imagesCount / self::IMAGE_LIMIT);
+        $io->progressStart($imagesCount);
+        for ($i = 0; $i < $iterations; ++$i) {
+            $imagesData = $this->imageGateway->getImagesData(
+                $i * self::IMAGE_LIMIT,
+                self::IMAGE_LIMIT
+            );
+
+            foreach ($imagesData as $imageData) {
+                $imagePathsToNormalize = $this->updateImagePathsToNormalize(
+                    $imageData,
+                    $imagePathsToNormalize
+                );
+
+                $io->progressAdvance();
+            }
+        }
+        $io->progressFinish();
+
+        return $imagePathsToNormalize;
+    }
+
+    protected function normalizeImagePaths(array $imagePathsToNormalize, SymfonyStyle $io): array
```

Comment:
```suggestion
    private function normalizeImagePaths(array $imagePathsToNormalize, SymfonyStyle $io): array
```

Additionally, this method can be changed to work as a `Generator`, and we could `yield` and remove images immediately after performing each & every `commit` operation here. Unless there is a reason for the removal to be postponed until the very end?

Reply:
This is postponed to make sure that the file is not used by two or more iterations.